### PR TITLE
E2E Testing Setup

### PR DESCRIPTION
**Pull request by kadey001**
Set up e2e testing with **playwright** and report generation with **Allure**
Added simple starter test for landing page

### CODE REVIEW

File: .github/workflows/development-workflow.yml [hunk 1]
```diff
@@ -6,7 +6,6 @@ on:
         types: [opened, edited, synchronize]
 
 env:
-    PROJECT_ID: ${{ secrets.GKE_PROJECT_ID }}
```
Comment (DevelopmentByDavid) [inline .github/workflows/development-workflow.yml:9]: Is there a reason this was deleted?

File: .github/workflows/development-workflow.yml [hunk 2]
```diff
@@ -15,6 +14,7 @@ env:
     SERVER_IMAGE: prytaneum-server
     NAMESPACE: development
     GRAPHQL_URL: https://dev.prytaneum.io/graphql
+    CI: 1
```
Comment (DevelopmentByDavid) [inline .github/workflows/development-workflow.yml:19]: Could you add a description of what this is for in a comment?

File: .github/workflows/development-workflow.yml [hunk 3]
```diff
@@ -58,20 +58,47 @@ jobs:
...
+              run: yarn install
+
+            - name: Prisma Generate
+              run: yarn workspace @app/server generate
+
+            - name: Install Playwright
+              run: yarn workspace @app/client run playwright install --with-deps
+
+            - name: Install Playwright Microsoft Edge
+              run: yarn workspace @app/client run playwright install msedge
```
Comment (DevelopmentByDavid) [inline .github/workflows/development-workflow.yml:88]: What's the motivation behind installing edge specifically here?  A comment would be helpful on why edge but not other browsers.

File: app/client/package.json [hunk 4]
```diff
@@ -9,7 +9,12 @@
         "build": "next build",
         "relay": "relay-compiler --watch",
         "typecheck": "tsc -p tsconfig.prod.json --noEmit",
-        "lint": "eslint \"src/**\""
+        "lint": "eslint \"src/**\"",
+        "test:e2e": "playwright test && yarn allure:generate-report",
+        "test:e2e:ci": "dotenv -v CI=1 -- playwright test && yarn allure:generate-report",
+        "test:e2e:headed": "playwright test --headed && yarn allure:generate-report",
+        "allure:generate-report": "allure generate ./allure-results --clean",
+        "allure:report": "allure open ./allure-report"
```
Comment (DevelopmentByDavid) [inline app/client/package.json]: I think it may be better to have a `ci:test` command, which will run all of the appropriate e2e tests and in the future also run the unit tests.

File: app/client/package.json [hunk 5]
```diff
@@ -9,7 +9,12 @@
         "build": "next build",
         "relay": "relay-compiler --watch",
         "typecheck": "tsc -p tsconfig.prod.json --noEmit",
-        "lint": "eslint \"src/**\""
+        "lint": "eslint \"src/**\"",
+        "test:e2e": "playwright test && yarn allure:generate-report",
+        "test:e2e:ci": "dotenv -v CI=1 -- playwright test && yarn allure:generate-report",
```
Comment (DevelopmentByDavid) [inline app/client/package.json]: This line could be `"test:e2e:ci": "dotenv -v CI=1 -- yarn run test:e2e"` I believe.

File: app/client/.gitignore [hunk 6]
```diff
@@ -32,4 +32,11 @@ yarn-error.log*
...
-_unused
\ No newline at end of file
+
+_unused
+
+# playwright
+test-results/
+playwright-report/
+allure-report/
+allure-results/
```
Comment (DevelopmentByDavid) [inline app/client/.gitignore:42]: Should these also be added to the .dockerignore?

File: app/client/playwright.config.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
+import type { PlaywrightTestConfig } from '@playwright/test';
```
Comment (DevelopmentByDavid) [inline app/client/playwright.config.ts:1]: Out of curiosity, why is this file ts?  Is that what the example did?  It's a bit odd to have a config file in typescript.

File: app/client/src/features/accounts/UserMenu/UserMenu.tsx [hunk 8]
```diff
@@ -179,10 +179,20 @@ export function UserMenu({ className, queryRef }: UserMenuProps) {
                     </Button>
                     <ResponsiveDialog open={type === 'login'} onClose={close}>
                         <DialogContent>
-                            <LoginForm onSuccess={() => { close(); router.reload(); }} />
+                            <LoginForm
+                                onSuccess={() => {
+                                    close();
+                                    router.reload();
+                                }}
+                            />
```
Comment (DevelopmentByDavid) [inline app/client/src/features/accounts/UserMenu/UserMenu.tsx:187]: What happened to the formatting here?

File: app/client/src/features/accounts/UserMenu/UserMenu.tsx [hunk 9]
```diff
@@ -179,10 +179,20 @@ export function UserMenu({ className, queryRef }: UserMenuProps) {
...
                         </DialogContent>
                     </ResponsiveDialog>
-                    <Button color='primary' variant='outlined' onClick={handleClick('register')}>
+                    <Button
+                        data-test-id='appbar-register-button'
```
Comment (DevelopmentByDavid) [inline app/client/src/features/accounts/UserMenu/UserMenu.tsx:191]: Does this get filtered out when transpiling to js for prod?

File: app/client/src/features/landing/CallToAction/CallToAction.tsx [hunk 10]
```diff
@@ -50,28 +50,39 @@ export function CallToAction() {
...
                 </Typography>
-                {user ? 
-                    <Button variant='contained' color='secondary' className={classes.button} onClick={() => router.push('/app/home')}>
+                {user ? (
+                    <Button
+                        variant='contained'
+                        color='secondary'
+                        className={classes.button}
+                        onClick={() => router.push('/app/home')}
+                    >
```
Comment (DevelopmentByDavid) [inline app/client/src/features/landing/CallToAction/CallToAction.tsx:67]: Formatting looks weird here

File: .husky/pre-commit [hunk 11]
```diff
@@ -11,3 +11,7 @@ printf "No type errors! 🎉🎉🎉\n"
 printf "Linting project...\n"
 yarn workspaces foreach -p run lint
 printf "No lint errors! 🎉🎉🎉\n"
+
+printf "E2E testing...\n"
+yarn workspace @app/client run test:e2e:ci
+printf "All tests passing! 🎉🎉🎉\n"
```
Comment (DevelopmentByDavid) [inline .husky/pre-commit]: How fast do these run btw?  If it's like 10 minutes, may be more worthwhile to run on CI/CD rather than making the dev wait for e2e tests to pass before commiting.

**Review by DevelopmentByDavid — CHANGES_REQUESTED** [review]
This is awesome!  Really exciting to see tests being worked on.  Thank you!

File: .github/workflows/development-workflow.yml (hunk 1 above)
Comment (kadey001) [reply .github/workflows/development-workflow.yml:9]: Since the deployment is in a different workflow now it's outdated and unnecessary.

File: .github/workflows/development-workflow.yml (hunk 2 above)
Comment (kadey001) [reply .github/workflows/development-workflow.yml:19]: Playwright uses this environment variable to set up some of the configuration differently in the github action. Using it we can get different more detailed output when running the tests locally but it can do a more simplified version when the action runs.

File: .github/workflows/development-workflow.yml (hunk 3 above)
Comment (kadey001) [reply .github/workflows/development-workflow.yml:88]: The other command installs all the other browsers but apparently it doesn't install edge as well so I added that to ensure it was also able to run the tests on edge.

File: app/client/package.json (hunk 4 above)
Comment (kadey001) [reply app/client/package.json]: We can just add that in when we do have those tests and have it call test:e2e as part of it though. All we'd have to do is update the command called in the workflow.

File: app/client/package.json (hunk 5 above)
Comment (kadey001) [reply app/client/package.json]: Yeah that might be a bit cleaner

File: app/client/.gitignore (hunk 6 above)
Comment (kadey001) [reply app/client/.gitignore:42]: Oh yes I think so, good catch.

File: app/client/playwright.config.ts (hunk 7 above)
Comment (kadey001) [reply app/client/playwright.config.ts:1]: I used their initialization and that part was set up by default. It gave an option for typescript so I chose it.

File: app/client/src/features/accounts/UserMenu/UserMenu.tsx (hunk 8 above)
Comment (kadey001) [reply app/client/src/features/accounts/UserMenu/UserMenu.tsx:187]: I added in a data-test-id on the register button to do the sample test so that extra section updated the formatting.

File: app/client/src/features/accounts/UserMenu/UserMenu.tsx (hunk 9 above)
Comment (kadey001) [reply app/client/src/features/accounts/UserMenu/UserMenu.tsx:191]: I'm not sure, I'd have to check. It was one of the suggestions by playwright but I might have to add something to next's config for it to filter it out.

File: .husky/pre-commit (hunk 11 above)
Comment (kadey001) [reply .husky/pre-commit]: Well it's relatively quick on my machine (like 30 seconds or so), but that's just one test and it may be slower on other's machines. I can remove it though if you think it's going to be a problem.

File: app/client/src/features/landing/CallToAction/CallToAction.tsx (hunk 10 above)
Comment (kadey001) [reply app/client/src/features/landing/CallToAction/CallToAction.tsx:67]: It was also a consequence of adding the data-test-id field. I ran prettier on it though so it should be the correct formatting.

File: .github/workflows/development-workflow.yml (hunk 2 above)
Comment (DevelopmentByDavid) [reply .github/workflows/development-workflow.yml:19]: That makes sense, please add that to a comment in the file.  Maybe even with a link to the docs where it explains that.

File: .github/workflows/development-workflow.yml (hunk 3 above)
Comment (DevelopmentByDavid) [reply .github/workflows/development-workflow.yml:88]: Makes sense.  Please add a comment describing that.

**Review by DevelopmentByDavid — APPROVED** [review]
looks good, please just add the comments explaining what the new things in the workflow are for

File: .husky/pre-commit (hunk 11 above)
Comment (DevelopmentByDavid) [reply .husky/pre-commit]: I think it's okay for now to leave it in tbh, but your call.  We can always add or remove it later.

File: app/client/src/features/landing/CallToAction/CallToAction.tsx (hunk 10 above)
Comment (DevelopmentByDavid) [reply app/client/src/features/landing/CallToAction/CallToAction.tsx:67]: Maybe it was github printing it weird.  It looks normal now.

Comment (DevelopmentByDavid) [reply app/client/src/features/landing/CallToAction/CallToAction.tsx:67]: Ah, yeah it was I see what happened now.  ignore the formatting comments

File: app/client/src/features/accounts/UserMenu/UserMenu.tsx (hunk 9 above)
Comment (DevelopmentByDavid) [reply app/client/src/features/accounts/UserMenu/UserMenu.tsx:191]: Yeah, it's not a big deal tbh, was more so curious.

Comment (DevelopmentByDavid) [reply app/client/src/features/accounts/UserMenu/UserMenu.tsx:191]: Mainly because, if there's a way to do it without having all of those artifacts in a production deployment, it's gonna be a pain to clean up once we have more tests.

Comment (DevelopmentByDavid) [reply app/client/src/features/accounts/UserMenu/UserMenu.tsx:191]: Maybe we can brainstorm more about different strategies for selecting particular buttons/actions.

File: app/client/playwright.config.ts (hunk 7 above)
Comment (DevelopmentByDavid) [reply app/client/playwright.config.ts:1]: Okay, makes sense.
